dipy/tracking/streamlinespeed.pyx

```diff
@@ -53,6 +53,63 @@ cdef void c_arclengths_from_arraysequence(Streamline points,
 
             arclengths[i] += sqrt(sum_dn_sqr)
 
+cdef void c_add_displacements(Streamline streamline,
+                              Streamline dField,
+                              Streamline new_streamline) nogil:
+
+    cdef:
+        np.npy_intp i
+
+    for i in range(streamline.shape[0]):
+        new_streamline[i, 0] = streamline[i, 0] + dField[i, 0]
+        new_streamline[i, 1] = streamline[i, 1] + dField[i, 1]
+        new_streamline[i, 2] = streamline[i, 2] + dField[i, 2]
+
+
+def add_displacements(streamlines, displacements):
+    ''' Adds warpfield displacements to streamlines
+
+    Parameters
+    ----------
+    streamlines : list of 2D ndarrays of shape[-1]==3
+    displacements : list of 2D ndarrays of shape[-1]==3
+
+    Returns
+    -------
+    new_streamlines : list of 2D ndarrays of shape[-1]==3
+
+    '''
+    new_streamlines = []
+    cdef np.npy_intp i
+    for i in range(len(streamlines)):
+        dtype = streamlines[i].dtype
+        # HACK: To avoid memleaks we have to recast with astype(dtype).
```

Comment:
Wow. Care to elaborate?

Reply:
Oh, I see - this is copied from other parts of streamlinespeed. Maybe someone else knows what this is about?

Reply:
yeah this was just copied from compress_streamlines.

Reply:
I don't know exactly why it produces memleaks without the `.astype` :-/ but it was causing issues with large tractogram (represented as a big list of ndarrays). I did make a series of tests check for that (e.g. https://github.com/nipy/dipy/blob/master/dipy/tracking/tests/test_streamline.py#L499). My guess is Cython is not releasing the refcount when it is done with `streamlines[i]`.